### PR TITLE
Message translate feature

### PR DESCRIPTION
**Pull request by Arseny271**

### CODE REVIEW

**Comment by RazaDroid** [comment]
"To Not Translate" option is missing

**Comment by RazaDroid** [comment]
If I'm selecting Any Language Other Than Application Language

> Only English in translated to that language
> Other languages are translated to English 

https://user-images.githubusercontent.com/37952910/232655229-3984c51e-1361-4826-9ab6-64f603975cdf.mp4

**Comment by Arseny271** [comment]
> If I'm selecting Any Language Other Than English (either application or do not translate)
> 
> > Only English in translated to that language
> > Other languages are translated to English
> 
> Like I select Hindi
> 
> English will translate to Hindi And other languages like Russian will translate to English

Try once to specify the translation language. App will remember your choice

**Comment by maryman0711** [comment]
вт, 11 апр. 2023 г., 22:36 Arseny271 ***@***.***>:

**Comment by tunxxxxx5555** [comment]
@stirpchat

File: app/src/main/java/org/thunderdog/challegram/Log.java [hunk 1]
```diff
@@ -407,7 +408,7 @@ private static void notifyLogFilesAltered () {
     TAG_YOUTUBE,
     TAG_CAMERA,
     TAG_EMOJI,
-
+    TAG_LANGUAGE,
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/Log.java:411]: From the 361 line of this file:
> // Don't forget to duplicate new values in log.h

`app/jni/log.h` and `app/jni/log.cpp` files must be updated when adding new log tags.

Also `version.jni` must be bumped after making some changes to native library.

File: app/src/main/java/org/thunderdog/challegram/unsorted/Settings.java [hunk 2]
```diff
@@ -1009,6 +1015,116 @@ private void upgradeSharedPreferences (SharedPreferences prefs, SharedPreference
...
+  public ArrayList<String> getTranslateLanguageRecents () {
+    String[] result = pmc.getStringArray(KEY_CHAT_TRANSLATE_RECENTS);
+    if (result != null) {
+      return new ArrayList<>(Arrays.asList(result));
+    }
+    return new ArrayList<>();
+  }
+
+  public void clearTranslateLanguageRecents () {
+    pmc.edit().remove(KEY_CHAT_TRANSLATE_RECENTS).apply();
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/unsorted/Settings.java:1125]:
```suggestion
    pmc.remove(KEY_CHAT_TRANSLATE_RECENTS);
```

File: app/src/main/java/org/thunderdog/challegram/util/LanguageDetector.java
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/util/LanguageDetector.java:1]: Code style: missing whitespaces before `(` in method declarations

**Review by phisikol — COMMENTED** [review]
https://user-images.githubusercontent.com/125762493/236197718-9630d1d1-697a-481c-8e99-eda8a89c13f7.mp4

**Review by Asman3316 — COMMENTED** [review]
مشترک۱۱۳ در تلگرام که ربات هست وعکس گربه دارد از طرف جتی هست یا خیر برسی شود

**Comment by Asman3316** [comment]
جواب من را خواهشن بدهد جتی تا بدانم مشکلی برایم پیش نیاید در امنیت کارم میخواهم بدانم گروه گربه وربات انداخل تلگرام از طرف جتی هست خواهشن جواب دهد جتی

**Comment by Asman3316** [comment]
> مشترک۱۱۳ در تلگرام که ربات هست وعکس گربه دارد از طرف جتی هست یا خیر برسی شود

File: app/src/main/java/org/thunderdog/challegram/component/chat/MessageView.java [hunk 3]
```diff
@@ -485,7 +485,7 @@ private static void selectMessage (MessagesController m, TGMessage msg, float to
   }
 
   private boolean onMessageClick (float x, float y) {
-    if (msg == null || msg instanceof TGMessageBotInfo) {
+    if (msg == null /*|| msg instanceof TGMessageBotInfo*/) {
```
[the remark this answers is not in the document]
Comment (payfazz99) [reply app/src/main/java/org/thunderdog/challegram/component/chat/MessageView.java:488]: - _[]()_

File: app/src/main/java/org/thunderdog/challegram/data/TGMessageFile.java [hunk 4]
```diff
@@ -695,4 +700,66 @@ protected boolean onLocaleChange () {
...
+      if (file.hasCaption()) {
+        resultText = Td.concat(resultText, sep, getTranslationSafeText(file.serverCaption));
+        filesWithCaption++;
+      } else {
+        resultText = Td.concat(resultText, sep);
+      }
+    }
+
+    return filesWithCaption > 0? Td.trim(resultText): null;
+  }
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/data/TGMessageFile.java:731]: There is a slight side effect of this solution, which seems to be present only in the "inline" translation mode, that all 📄 emoji in captions to media files get replaced with 📑 emoji, which itself isn't too critical, but it's there even if it's just 1 file, not an album, in which case it seems `TranslateMessageText` method could have been used instead.

**Review by vkryl — APPROVED** [review]
No critical issues found by code review, only minor / code style improvements could be made which can be done after merging.

**Comment by GaiaLoveWarrior** [comment]
Haiii gais ❤️ Blessings worldwide 🗺️🎐 I'd love to Co-Create a secure KEYBOARD + CLIPBOARD with Telegram's Devs! t.key or t.ark (like Noah's ark)
Your principles are admirable and I value so so much how brave you Telegram Founders are, as true warriors you stand up to everyone to keep user's safety. I love you dearly for leading new generations wisely. So please create a KEYBOARD and +3 ply Garden size CLIPBOARD (we can download from Play Store) to keep what we type, screenshot and copy paste private. 
(I'm a 43 yr autistic from 🇨🇷 aiming to learn to code with the best, ie YOU 😁 My apologies for writing here about a new topic, I'm learning how to use this site just to communicate with you 💗 Stay blessed sweethearts Your work and efforts are priceless and I'm so grateful for your values and principles just had to say it.)
 Hugs 
Cristina

**Pull request merged**